Experimental porting to QOpenGLWidget

mgian opened the pull request:
This should be a porting to the new QOpenGLWidget. It seems to work pretty well.

lrineau commented:
@maxGimeno: Can you have a look at this pull-request and comment? I know it was something that was wanted, for the port to Android.

maxGimeno commented:
The only problem I see is that the doc has not been updated and still cites QGLWidget everywhere instead of QOpenGLWidget. I also wonder why I had to use QOpenGLFunctions and it is not done here, but I guess it is because we wanted OpenGL ES compatible code, which is not an issue here.

GillesDebunne commented on QGLViewer/qglviewer.h:
```diff
@@ -557,6 +561,10 @@ public Q_SLOTS:
 	/*! Same as bufferTextureMaxU(), but for the v texture coordinate. */
 	qreal bufferTextureMaxV() const { return bufferTextureMaxV_; }
+#if (QT_VERSION >= QT_VERSION_CHECK(5, 4, 0))
+    void renderText(double x, double y, const QString &str, const QFont &font);
```
This is not exactly the signature in QGLWidget , which expects `int`s
Note that for `renderText(int x, int y,...)`, _x and y are specified in window coordinates_
while `renderText(double x, double y, double z...) _x, y and z are specified in scene or object coordinates relative to the currently set projection and model matrices_

GillesDebunne commented on QGLViewer/qglviewer.h:
```diff
@@ -160,13 +161,16 @@ public Q_SLOTS:
-	void setBackgroundColor(const QColor& color) { backgroundColor_=color; qglClearColor(color); }
-	/*! Sets the foregroundColor() of the viewer, used to draw visual hints. See also setBackgroundColor(). */
+        setForegroundColor(). */
+#if (QT_VERSION >= QT_VERSION_CHECK(5, 4, 0))
```
Why not use glClearColor in all cases and skip the `#if`?

GillesDebunne commented on QGLViewer/qglviewer.cpp:
```diff
@@ -342,7 +344,12 @@ void QGLViewer::postDraw()
 
 	glDisable(GL_COLOR_MATERIAL);
-	qglColor(foregroundColor());
+#if (QT_VERSION >= QT_VERSION_CHECK(5, 4, 0))
```
See comment below. Can we skip this test to make it simpler?

GillesDebunne commented on QGLViewer/qglviewer.cpp:
```diff
@@ -685,6 +692,30 @@ void QGLViewer::drawLight(GLenum light, qreal scale) const
+    // Identify x and y locations to render text within widget
+    int height = this->height();
+    GLdouble textPosX = 0, textPosY = 0, textPosZ = 0;
+    //project(x, y, 0.0, &textPosX, &textPosY, &textPosZ);
```
This commented line probably comes from the difference between the x,y and the x,y,z versions of this method. See comment below.
I believe you need to provide both, since both are used, with signatures similar to the original API.

GillesDebunne commented on examples/contribs/dvonn/drawer.cpp:
```diff
@@ -373,7 +373,7 @@ Drawer::drawStatus(const Board::ConstStackHandle& s,QGLViewer* v) const
   translateTo(s.stackCoord(),s->height()*pieceH+pieceH/2.0f);
   glColor3f(1.0f,1.0f,0.0f);
-  v->renderText(0.5f*caseD,0.5f*caseD,0.0f,QString("%1").arg(s.stackStatus()));
+  v->renderText(0.5f*caseD,0.5f*caseD,QString("%1").arg(s.stackStatus()), QFont());
```
QFont() not strictly needed here. Note this is using the x,y,z (i.e. space coordinates) version of `renderText`, which is not what you implemented.

GillesDebunne commented on examples/contribs/terrain/viewer.cpp:
```diff
@@ -324,7 +324,7 @@ QString Viewer::helpString() const
   text += "Press <b>X</b> to switch texturing on and off.<br>";
   text += "Press <b>O</b> to switch shading on and off.<br>";
-  text += "Press <b>L</b> to cycle through different light source positions (+45°).<br>";
+  text += "Press <b>L</b> to cycle through different light source positions (+45??).<br>";
```
Replace by 'degrees' if there's an issue with this character.

mgian commented:
Removing the #ifdef block as the side effect, as I understand, to make it incompatible with earlier versions of the library, which as today are at least 3 years old, since the new widget is shipped starting with this version of Qt. I am not sure it is a good idea to drop the support fot Qt libraries < 5.4.

I will fix the degrees typo and the renderText and update the pull request

GillesDebunne commented:
Thanks for these.

My understanding is that `qglColor` is a simple helper function that is not present in `QOpenGLWidget`. But I would rather replace it everywhere by a native OpenGL `glColor4f` call, rather than keep `#ifdef` around. This is already what you did line 153 in `examples/constrainedCamera/constrainedCamera.cpp`. I don't think there's a backward compatibility issue with this approach.

mgian commented:
I updated the pull request, but the reimplementation of renderText() method seems to be a little more difficult, so I need some more time.

oracle3001 commented:
You can do renderText like this

```
void QGLViewer::renderText(int textPosX, int textPosY, const QString &text, const QFont & font, const QColor& fontColor) {

	// Render text
	QPainter painter(this);
	painter.setPen(fontColor);
	painter.setFont(font);
	painter.drawText(textPosX, textPosY, text);
	painter.end();
	
}
```

GillesDebunne commented:
Interesting workaround, thanks. Needs to be fully tested, but that's a good
option.

On Fri, Jun 9, 2017 at 12:23 PM, oracle3001 <notifications@github.com>
wrote:

> You can do renderText like this
>
> void QGLViewer::renderText(int textPosX, int textPosY, const QString &text, const QFont & font, const QColor& fontColor) {
>
> 	// Render text
> 	QPainter painter(this);
> 	painter.setPen(fontColor);
> 	painter.setFont(font);
> 	painter.drawText(textPosX, textPosY, text);
> 	painter.end();
> 	
> }
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/GillesDebunne/libQGLViewer/pull/36#issuecomment-307353640>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABzymjh9RJbuSJ7vL3RrooxefU6D0l47ks5sCR0RgaJpZM4NAq7w>
> .
>

mgian commented:
I updated the my code using the workaround suggested by @oracle3001 and it seems to work reasonably well. I have also updated the documentation to refer to the new QOpenGLWidget
However I have not update the examples in the 'contrib'

GillesDebunne commented:
I manually merged your branch into master, and added a few extra changes.

Thanks a lot for this contribution. Closing this PR.

G
